### ISSUE TITLE
Need a way to propogate job cancelation to a cancelable coroutine

### ISSUE THREAD

**Issue by CodingDoug** [Kotlin/kotlin-coroutines#108]
I'm thinking about implementing coroutine support an existing asynchronous API.  Everything is great when I use a `suspendCancellableCoroutine`.  I can call `resume()` and `resumeWithException()` on the provided `CancellableContinuation` in the callbacks from the async API.  So, if the consumer of my API wants to use coroutines, it's easy to provide a coroutine suspend `await()` wrapper via an extension function.

However, if my async API is itself cancelable, I currently have no way (that I can see) of propagating the API consumer's call to `Job.cancel()` to my API.  Without this, my API becomes much less compelling to consume via coroutines.  I would love to see some way for a `CancellableContinuation` to provide a hook I can use to know if the developer tried to cancel the Job, and allow me to react to that.

**Comment by JakeWharton** [comment]
You can pass a lambda to `invokeOnCompletion` which checks the `isCanceled` flag to perform a cancelation action.

**Comment by elizarov** [comment]
@CodingDoug Can you, please, provide some kind of self-contained example that shows your problem and submit it as a bug to https://github.com/Kotlin/kotlinx.coroutines/issues

**Comment by CodingDoug** [comment]
@elizarov I figured out what was going wrong.  I was calling a `suspend fun` that returned a `suspendCancellableCoroutine` instead of just calling a `suspendCancellableCoroutine` fun itself.  I had to dig around existing code for a demo of this, as the existing docs weren't quite clear enough at first read (I'm only moderately experienced in Kotlin - understanding things from docs alone may have required more expertise).
